File: src/main/java/com/koldyr/csv/processor/BatchDBProcessor.java
package com.koldyr.csv.processor;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Collection;
import java.util.StringJoiner;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;

import org.slf4j.LoggerFactory;

import com.koldyr.csv.model.PoolType;
import com.koldyr.csv.model.ProcessorContext;

/**
 * Description of class DBProcessor
 *
 * @created: 2018.03.03
 */
public abstract class BatchDBProcessor implements Callable<Object> {

    protected final DecimalFormat format;
    protected final ProcessorContext context;

    protected BatchDBProcessor(ProcessorContext context) {
        this.context = context;

        final DecimalFormatSymbols decimalFormatSymbols = new DecimalFormatSymbols();
        decimalFormatSymbols.setGroupingSeparator(',');
        format = new DecimalFormat("###,###,###,###", decimalFormatSymbols);
    }

    @Override
    public Object call() {
        String tableName = context.getNextTable();
        while (tableName != null) {
            processTable(tableName);
            tableName = context.getNextTable();
        }

        return null;
    }

    protected abstract void processTable(String tableName);

    protected void checkResults(String tableName, int pageCount, Collection<Future<Integer>> results) throws InterruptedException {
        final int count = results.stream().mapToInt(result -> {
            try {
                return result.get();
            } catch (Exception e) {
                return 0;
            }
        }).sum();

        LoggerFactory.getLogger(getClass()).debug("checkResults - Expected: {}, Actual: {}", pageCount, count);

        if (count != pageCount) {
            throw new InterruptedException("Error exporting " + tableName);
        }
    }

    protected long getRowCount(Connection connection, String tableName) throws SQLException {
        long rowCount = 0;

        ResultSet resultSet = null;
        try (Statement statement = connection.createStatement()) {
            resultSet = statement.executeQuery("SELECT count(1) FROM " + context.getSchema() + '.' + tableName);

            if (resultSet.next()) {
                rowCount = resultSet.getLong(1);
            }
        } finally {
            if (resultSet != null) {
                resultSet.close();
            }
        }
        return rowCount;
    }

    protected void release(Connection connection, PoolType type) {
        if (connection != null) {
            try {
                context.release(type, connection);
            } catch (Exception e) {
                LoggerFactory.getLogger(getClass()).error(e.getMessage(), e);
            }
        }
    }

    public static String createInsertSql(String schema, String tableName, int columnCount) {
        StringJoiner values = new StringJoiner(",");
        for (int columnIndex = 1; columnIndex <= columnCount; columnIndex++) {
            values.add("?");
        }

        return "INSERT INTO \"" + schema + "\".\"" + tableName + "\" VALUES (" + values + ')';
    }
}

File: src/main/kotlin/com/koldyr/csv/processor/BatchDBProcessor.kt
package com.koldyr.csv.processor

import com.koldyr.csv.db.SQLStatementFactory
import com.koldyr.csv.model.PoolType
import com.koldyr.csv.model.ProcessorContext
import org.slf4j.LoggerFactory
import java.sql.Connection
import java.sql.SQLException
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.concurrent.Callable
import java.util.concurrent.Future

/**
 * Description of class DBProcessor
 *
 * @created: 2018.03.03
 */
abstract class BatchDBProcessor
    protected constructor(protected val context: ProcessorContext) : Callable<Any> {

    protected val format: DecimalFormat

    init {
        val decimalFormatSymbols = DecimalFormatSymbols()
        decimalFormatSymbols.groupingSeparator = ','
        format = DecimalFormat("###,###,###,###", decimalFormatSymbols)
    }

    override fun call(): Any? {
        var tableName: String? = context.nextTable
        while (tableName != null) {
            processTable(tableName)
            tableName = context.nextTable
        }

        return null
    }

    protected abstract fun processTable(tableName: String)

    @Throws(InterruptedException::class)
    protected fun checkResults(tableName: String, pageCount: Int, results: Collection<Future<Int>>) {
        val count = results.stream().mapToInt { result ->
            return@mapToInt try {
                result.get()
            } catch (e: Exception) {
                0
            }
        }.sum()

        LoggerFactory.getLogger(javaClass).debug("checkResults - Expected: {}, Actual: {}", pageCount, count)

        if (count != pageCount) {
            throw InterruptedException("Error exporting $tableName")
        }
    }

    @Throws(SQLException::class)
    protected fun getRowCount(connection: Connection, tableName: String): Long {
        var rowCount: Long = 0

        connection.createStatement().use { statement ->
            val getRowCount = SQLStatementFactory.getRowCount(connection, context.srcSchema, tableName)
            statement.executeQuery(getRowCount).use {
                if (it.next()) {
                    rowCount = it.getLong(1)
                }
            }
        }
        return rowCount
    }

    protected fun release(connection: Connection?, type: PoolType) {
        if (connection != null) {
            try {
                context.release(type, connection)
            } catch (e: Exception) {
                LoggerFactory.getLogger(javaClass).error(e.message, e)
            }
        }
    }
}
